src/api/errors/api_error.rs

```diff
@@ -0,0 +1,89 @@
+use std::fmt;
+
+#[derive(Debug, thiserror::Error)]
+pub struct ApiError {
+    inner: ApiErrorKind,
+    #[source]
+    source: Option<anyhow::Error>,
+}
+
+/// Represents API errors.
+#[derive(Copy, Clone, Eq, PartialEq, Debug, thiserror::Error)]
+pub(in crate::api) enum ApiErrorKind {
+    #[error("could not serialize value as JSON")]
+    CannotSerializeAsJson,
+    #[error("could not parse JSON response")]
+    BadJson,
+    #[error("not a JSON response")]
+    NotJson,
+    #[error("request failed because API URL was incorrectly formatted")]
+    BadApiUrl,
+    #[error("organization not found")]
+    OrganizationNotFound,
+    #[error("resource not found")]
+    ResourceNotFound,
+    #[error("Project not found. Please check that you entered the project and organization slugs correctly.")]
+    ProjectNotFound,
+    #[error("release not found")]
+    ReleaseNotFound,
+    #[error("chunk upload endpoint not supported by sentry server")]
+    ChunkUploadNotSupported,
+    #[error("API request failed")]
+    RequestFailed,
+    #[error("could not compress data")]
+    CompressionFailed,
+    #[error("region overrides cannot be applied to absolute urls")]
+    InvalidRegionRequest,
+    #[error(
+        "Auth token is required for this request. Please run `sentry-cli login` and try again!"
+    )]
+    AuthMissing,
+}
+
+impl fmt::Display for ApiError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.inner, f)
+    }
+}
+
+impl ApiError {
+    pub(in crate::api) fn with_source<E>(kind: ApiErrorKind, source: E) -> ApiError
+    where
+        E: Into<anyhow::Error>,
+    {
+        ApiError {
+            inner: kind,
+            source: Some(source.into()),
+        }
+    }
+
+    pub(in crate::api) fn kind(&self) -> ApiErrorKind {
```

Comment:
interesting, first time I see this syntax. Is that the same as `pub(super)`? Hm, actually that would be equivalent to `crate::api::errors` I believe.

Reply:
No, actually `pub(super)` is syntactical sugar for `pub(in super)`. `pub(in xyz)` restricts visibility to module `xyz` and `xyz`'s submodules ([see docs](https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself)).

In this case, we need to use `pub(in crate::api)` rather than `pub(super)`, since this function gets called in the `src/api/mod.rs`.